### PR TITLE
add a heatmap printer

### LINKED ISSUE #472: Heatmap printer
I want to add a heatmap printer for pterm where you can customize the colors for the range of numbers.
The printer should have the option to use RGB values but by default it should use the colors supported by every terminal.

### PR DESCRIPTION
**Pull request by floaust**
### Description
Added a heatmap printer.

- [ ] Bug Fix
- [x] New Feature
- [ ] Documentation
- [ ] Other

### Related Issue
Fixes #472 

### To-Do Checklist
- [x] I tested my changes
- [x] I have commented every method that I created/changed
- [x] I updated the examples to fit with my changes
- [x] I have added tests for my newly created methods

### CODE REVIEW

**Comment by floaust** [comment]
- [ ] tests need to be added
- [ ] docs need to be added
- [ ] demo needs to be refactored and split into multiple demos

**Comment by floaust** [comment]
@KarolosLykos I have merged your tests into this branch, now the code for the heatmap printer is more or less final, if you want you can add even more tests. :)

**Comment by MarvinJWendt** [comment]
No more tests are needed, we need to refactor the testing system anyway. The tests @KarolosLykos added are enough for now :)

File: heatmap_printer.go [hunk 1]
```diff
@@ -0,0 +1,657 @@
...
+				}
+				buffer.WriteString(p.AxisStyle.Sprint(ct))
+				if p.Grid {
+					buffer.WriteString(p.SeparatorStyle.Sprint(p.VerticalSeparator))
+				}
+			}
+			var ct string
+			if p.OnlyColoredCells {
+				ct = internal.CenterText(" ", lineLength)
+				if len(ct) < lineLength {
```
Comment (MarvinJWendt) [inline heatmap_printer.go:293]: 🚫 **[golangci]** <sub>reported by [reviewdog](https://github.com/reviewdog/reviewdog) :dog:</sub><br>SA9003: empty branch (staticcheck)

**Comment by sean-** [comment]
What's the current status of this PR?

**Comment by sean-** [comment]
Adding the sample output so reviewers of the PR can see this PRs utility:
image
image
image
image
image
image
image
This is a handy printer for those of us who live in the world of histograms.

**Comment by floaust** [comment]
Hi @sean-, working on it again. Will be out soon. Just needed to do some bug fixing, testing, refactorig and adding of a few features.

File: _examples/heatmap/custom_colors/main.go [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+		{0.9, -0.5, -0.1, 0.3, 1, -0.7, -0.3, 0.1, 0.7, -0.9, -0.5, 0.2, 0.6},
+		{0.5, 0.6, 0.1, -0.2, -0.7, 0.8, 0.6, 0.1, -0.5, -0.7, 0.7, 0.3, 0.0},
+	}
+
+	headerData := pterm.HeatmapAxis{
+		XAxis: []string{"a", "b", "c", "d", "e", "f", "g", "h", "i", "j", "k", "l", "m"},
+		YAxis: []string{"1", "2", "3", "4", "5"},
+	}
+
+	pterm.Println("The following table has no rgb (supported by every terminal), no axis data and a legend.\n")
```
Comment (MarvinJWendt) [inline _examples/heatmap/custom_colors/main.go:21]:
```suggestion
	pterm.Info.Println("The following table has no rgb (supported by every terminal), no axis data and a legend.")
	pterm.Println()
```

File: _examples/heatmap/custom_legend/main.go [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+		{0.9, -0.5, -0.1, 0.3, 1, -0.7, -0.3, 0.1, 0.7, -0.9, -0.5, 0.2, 0.6},
+		{0.5, 0.6, 0.1, -0.2, -0.7, 0.8, 0.6, 0.1, -0.5, -0.7, 0.7, 0.3, 0.0},
+	}
+
+	headerData := pterm.HeatmapAxis{
+		XAxis: []string{"a", "b", "c", "d", "e", "f", "g", "h", "i", "j", "k", "l", "m"},
+		YAxis: []string{"1", "2", "3", "4", "5"},
+	}
+
+	pterm.Println("The following table has rgb (not supported by every terminal), axis data and a custom legend.\n")
```
Comment (MarvinJWendt) [inline _examples/heatmap/custom_legend/main.go:21]:
```suggestion
	pterm.Info.Println("The following table has rgb (not supported by every terminal), axis data and a custom legend.")
	pterm.Println()
```

File: _examples/heatmap/custom_rgb/main.go [hunk 4]
```diff
@@ -0,0 +1,25 @@
...
+		{0.9, -0.5, -0.1, 0.3, 1, -0.7, -0.3, 0.1, 0.7, -0.9, -0.5, 0.2, 0.6},
+		{0.5, 0.6, 0.1, -0.2, -0.7, 0.8, 0.6, 0.1, -0.5, -0.7, 0.7, 0.3, 0.0},
+	}
+
+	headerData := pterm.HeatmapAxis{
+		XAxis: []string{"a", "b", "c", "d", "e", "f", "g", "h", "i", "j", "k", "l", "m"},
+		YAxis: []string{"1", "2", "3", "4", "5"},
+	}
+
+	pterm.Println("The following table has rgb (not supported by every terminal), axis data and a legend.\n")
```
Comment (MarvinJWendt) [inline _examples/heatmap/custom_rgb/main.go:21]:
```suggestion
	pterm.Info.Println("The following table has rgb (not supported by every terminal), axis data and a legend.")
	pterm.Println()
```

File: _examples/heatmap/no_grid/main.go [hunk 5]
```diff
@@ -0,0 +1,25 @@
...
+		{0.9, -0.5, -0.1, 0.3, 1, -0.7, -0.3, 0.1, 0.7, -0.9, -0.5, 0.2, 0.6},
+		{0.5, 0.6, 0.1, -0.2, -0.7, 0.8, 0.6, 0.1, -0.5, -0.7, 0.7, 0.3, 0.0},
+	}
+
+	headerData := pterm.HeatmapAxis{
+		XAxis: []string{"a", "b", "c", "d", "e", "f", "g", "h", "i", "j", "k", "l", "m"},
+		YAxis: []string{"1", "2", "3", "4", "5"},
+	}
+
+	pterm.Println("The following table has rgb (not supported by every terminal), axis data and a legend.\n")
```
Comment (MarvinJWendt) [inline _examples/heatmap/no_grid/main.go:21]:
```suggestion
	pterm.Info.Println("The following table has rgb (not supported by every terminal), axis data and a legend.")
	pterm.Println()
```

File: _examples/heatmap/separated/main.go [hunk 6]
```diff
@@ -0,0 +1,23 @@
...
+		{0.9, -0.5, -0.1, 0.3, 1, -0.7, -0.3, 0.1, 0.7, -0.9, -0.5, 0.2, 0.6},
+		{0.5, 0.6, 0.1, -0.2, -0.7, 0.8, 0.6, 0.1, -0.5, -0.7, 0.7, 0.3, 0.0},
+	}
+
+	headerData := pterm.HeatmapAxis{
+		XAxis: []string{"a", "b", "c", "d", "e", "f", "g", "h", "i", "j", "k", "l", "m"},
+		YAxis: []string{"1", "2", "3", "4", "5"},
+	}
+
+	pterm.Println("The following table has no rgb (supported by every terminal), no axis data and no legend.\n")
```
Comment (MarvinJWendt) [inline _examples/heatmap/separated/main.go:19]:
```suggestion
	pterm.Info.Println("The following table has no rgb (supported by every terminal), no axis data and no legend.")
	pterm.Println()
```

File: _examples/heatmap/demo/main.go [hunk 7]
```diff
@@ -0,0 +1,23 @@
...
+		{0.9, -0.5, -0.1, 0.3, 1, -0.7, -0.3, 0.1, 0.7, -0.9, -0.5, 0.2, 0.6},
+		{0.5, 0.6, 0.1, -0.2, -0.7, 0.8, 0.6, 0.1, -0.5, -0.7, 0.7, 0.3, 0.0},
+	}
+
+	headerData := pterm.HeatmapAxis{
+		XAxis: []string{"a", "b", "c", "d", "e", "f", "g", "h", "i", "j", "k", "l", "m"},
+		YAxis: []string{"1", "2", "3", "4", "5"},
+	}
+
+	table := pterm.DefaultHeatmap.WithAxisData(headerData).WithData(data)
```
Comment (MarvinJWendt) [inline _examples/heatmap/demo/main.go:21]: The Demo should be the "nicest" version of the printer. Could you enable RGB here?

File: heatmap_printer.go [hunk 8]
```diff
@@ -0,0 +1,807 @@
...
+	TextRGB                    RGB
+	Writer                     io.Writer
+
+	minValue float32
+	maxValue float32
+
+	rgbLegendValue int
+}
+
+var colorComplement = map[Color]Color{
```
Comment (MarvinJWendt) [inline heatmap_printer.go:89]: Maybe rename to `complementaryColors`?

File: heatmap_printer.go [hunk 9]
```diff
@@ -0,0 +1,807 @@
...
+	}
+	return &p
+}
+
+// WithGrid returns a new HeatmapPrinter with a grid.
+func (p HeatmapPrinter) WithGrid(b ...bool) *HeatmapPrinter {
+	b2 := internal.WithBoolean(b)
+	p.Grid = b2
+	if !b2 && p.Boxed {
+		p.Boxed = false
```
Comment (MarvinJWendt) [inline heatmap_printer.go:164]: Please include this behavior in the doc comment.

File: heatmap_printer.go [hunk 10]
```diff
@@ -0,0 +1,807 @@
...
+	p.OnlyColoredCells = b2
+	return &p
+}
+
+// WithLegendOnlyColoredCells returns a new HeatmapPrinter with legend with only colored cells.
+func (p HeatmapPrinter) WithLegendOnlyColoredCells(b ...bool) *HeatmapPrinter {
+	b2 := internal.WithBoolean(b)
+	p.LegendOnlyColoredCells = b2
+	if b2 {
+		p.Legend = true
```
Comment (MarvinJWendt) [inline heatmap_printer.go:187]: Please include this behavior in the doc comment.

File: heatmap_printer.go [hunk 11]
```diff
@@ -0,0 +1,807 @@
...
+// WithTCrossSeparator returns a new HeatmapPrinter with a specific TCrossSeparator.
+func (p HeatmapPrinter) WithTCrossSeparator(s string) *HeatmapPrinter {
+	p.TCrossSeparator = s
+	return &p
+}
+
+// WithRGBRange returns a new HeatmapPrinter with a specific RGBRange.
+func (p HeatmapPrinter) WithRGBRange(rgb ...RGB) *HeatmapPrinter {
+	p.RGBRange = rgb
+	p.EnableRGB = true
```
Comment (MarvinJWendt) [inline heatmap_printer.go:288]: This function should not enable RGB, as this might be unexpected.

File: heatmap_printer.go [hunk 12]
```diff
@@ -0,0 +1,807 @@
...
+	return &p
+}
+
+// WithColors returns a new HeatmapPrinter with a specific Colors.
+func (p HeatmapPrinter) WithColors(colors ...Color) *HeatmapPrinter {
+	p.Colors = colors
+	p.EnableRGB = false
```
Comment (MarvinJWendt) [inline heatmap_printer.go:295]: This function should not disable RGB, as this might be unexpected.

File: heatmap_printer.go [hunk 13]
```diff
@@ -0,0 +1,807 @@
...
+	if err := p.errCheck(); err != nil {
+		return "", err
+	}
+
+	if p.SeparatorStyle == nil {
+		p.SeparatorStyle = NewStyle()
+	}
+	if p.AxisStyle == nil {
+		p.AxisStyle = NewStyle()
+	}
```
Comment (MarvinJWendt) [inline heatmap_printer.go:316]: You can make use of Theme values here, rather than empty styles.

**Review by MarvinJWendt — CHANGES_REQUESTED** [review]

File: heatmap_printer.go (hunk 8 above)
Comment (floaust) [reply heatmap_printer.go:89]: we talked about this a few months and decided on this naming

File: heatmap_printer.go (hunk 10 above)
Comment (floaust) [reply heatmap_printer.go:192]: done

File: heatmap_printer.go (hunk 9 above)
Comment (floaust) [reply heatmap_printer.go:168]: done

**Review by MarvinJWendt — APPROVED** [review]
LGTM! 🚀

**Pull request merged**